docs/cpp_docs.rst

```diff
@@ -26,94 +44,556 @@ The beginning of both header and source code files should include
+
+1. In what stack version the object was deprecated, and when it will be removed.
+2. Reason for deprecation if this is useful information (e.g., object is superseded, duplicates functionality found elsewhere, etc.).
+3. New recommended way of obtaining the same functionality.
+
+.. code-block:: cpp
+
+   /**
+    * @deprecated Deprecated in 11_0. `ndobj_old` will be removed in 12_0; it
+    *     is replaced by `ndobj_new` because the latter works also with array
+    *     subclasses.
+    */
+
+.. _cpp-doxygen-extended-summary:
+
+Extended Summary
+----------------
+
+A few sentences giving an extended description.
+This section should be used to clarify *functionality*, not to discuss implementation detail or background theory, which should rather be explored in the :ref:`cpp-doxygen-notes` section below.
+You may refer to the parameters and the function name, but parameter descriptions still belong in the :ref:`cpp-doxygen-parameters` section unless they are very lengthy.
+
+.. _cpp-doxygen-tparameters:
+
+Template Parameters
+-------------------
+
+A series of ``@tparam`` tags, usually one for each template parameter. Each tag should have a description following the parameter name. You do *not* usually need to document default values; Doxygen will provide the default automatically. If the description extends over multiple lines, each line after the first must be indented.
+
+Parameters should be listed in the same order as they appear in the class, function, or method signature. 
+
+.. code-block:: cpp
+
+   /**
+    * Storage for arbitrary data with log(N) lookup.
     *
-    * @brief This message displayed in Doxygen Files index
+    * ...
     *
-    * @ingroup PackageName
-    * (Note: this needs exactly one @defgroup somewhere)
+    * @tparam T the type of data stored in the table
+    * @tparam ComparatorT callable defining a strict weak ordering for objects
+    *     of type `T`. Its `operator()` must accept two `T` and return `true`
+    *     if and only if the first argument comes before the second. It must
+    *     not throw exceptions.
+    */
+   template <typename T, typename ComparatorT = std::less<T>>
+   class LookupTable
+   {
+       ...
+   }
+
+When two or more consecutive template parameters have *exactly* the same description, they can be combined:
+
+.. code-block:: cpp
+
+   /**
+    * @param T,U the types of the pair components
+    */
+
+Do not use a space between the parameters.
+
+.. _cpp-doxygen-parameters:
+
+Function/Method Parameters
+--------------------------
+
+A series of ``@param`` tags, usually one for each parameter. Each tag should have a description following the parameter name. You do *not* usually need to document default arguments; Doxygen will provide the default automatically. If the description extends over multiple lines, each line after the first must be indented.
```

Comment:
Should specify how it should be indented.  Example below suggests additional should line up with the parameter name.  Note that we more frequently line up with the start of the description, which is cleaner but can lead to many more line breaks when parameter names are long.  I'm ambivalent about which one we go with.